<!-- ... -->
<launch>
  <arg name="rviz" default="false" />

  <arg name="pi/2" value="1.5707963267948966" />
  <arg name="angle" value="2.09" />
  <arg name="optical_rotate" value="0 0.06 0.85 -$(arg pi/2) 0 -$(arg angle)" />

  <node pkg="ocams" type="ocams" name="ocams_node" output="screen">
    <param name="resolution"    value="2"/> <!-- 0: 1280x960, 1: 1280x720, 2: 640x480, 3: 640x360 -->
    <param name="frame_rate"    value="15"/>
    <param name="exposure"      value="100"/>
    <param name="gain"          value="50"/>
    <param name="wb_blue"       value="200"/>
    <param name="wb_red"        value="150"/>
    <param name="auto_exposure" value="true"/>
    <param name="show_image"    value="false"/>

    <param name="left_frame_id" value="left_frame"/>
    <param name="right_frame_id" value="right_frame"/>
  </node>

  <node pkg="tf" type="static_transform_publisher" name="ocam_base_link1"
    args="$(arg optical_rotate) base_link left_frame 100" />
  <node pkg="tf" type="static_transform_publisher" name="ocam_base_link2"
    args="$0 -0.06 1.0 -$(arg pi/2) 0 -$(arg angle) base_link right_frame 100" />

  <group ns="/stereo" >
    <node pkg="nodelet" type="nodelet" name="stereo_nodelet"  args="manager"/>

    <node pkg="stereo_image_proc" type="stereo_image_proc" name="stereo_image_proc">
      <param name="disparity_range" value="96"/>
      <param name="min_disparity" value="0"/>
      <param name="speckle_size" value="1000"/>
    </node>

    <node pkg="nodelet" type="nodelet" name="disparity2cloud" args="load rtabmap_ros/point_cloud_xyz stereo_nodelet">
      <remap from="disparity/image"       to="disparity"/>
      <remap from="disparity/camera_info" to="left/camera_info"/>
      <remap from="cloud"                 to="cloudXYZ"/>
      <param name="voxel_size" type="double" value="0.02"/>
      <param name="decimation" type="int" value="2"/>
      <param name="max_depth"  type="double" value="3"/>
    </node>

  </group>

  <!-- elevation mapping -->
  <!-- Elevation mapping node -->
  <node pkg="elevation_mapping" type="elevation_mapping" name="elevation_mapping" output="screen">
    <rosparam command="load" file="$(find elevation_mapping)/config/silver_carriage.yaml" />
  </node>

  <!-- grid_map_visualization -->

  <node pkg="grid_map_visualization" type="grid_map_visualization" name="elevation_map_raw_visualization" output="screen">
    <param name="grid_map_topic" value="/elevation_mapping/elevation_map_raw" />
    <rosparam command="load" file="$(find elevation_mapping)/config/gridmap.yaml" />
  </node>

   <!-- Visualisation RVIZ --> 
   <node if="$(arg rviz)" pkg="rviz" type="rviz" name="rviz" args="-d $(find ocams)/config/rviz/grid.rviz"/>

   <!-- ROSBridge TCP Server -->
  <node pkg="rosbridge_server" type="rosbridge_tcp" name="rosbridge_tcp"  output="screen"/>

</launch>
